src/Microsoft.ML.Recommender/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,496 @@
+        }
+
+        /// <summary>
+        /// Create the mapper required by matrix factorization's predictor. That mapper maps two
+        /// index inputs (e.g., row index and column index) to the value located by the two indexes
+        /// in the training matrix. In recommender system where the training matrix stores ratings
+        /// from users to items, the mappers maps user ID and item ID to the rating of that item given
+        /// by the user.
+        /// </summary>
+        public ValueMapper<TXIn, TYIn, TOut> GetMapper<TXIn, TYIn, TOut>()
+        {
+            string msg = null;
+            msg = "Invalid TXIn in GetMapper: " + typeof(TXIn);
+            _host.Check(typeof(TXIn) == typeof(uint), msg);
+
+            msg = "Invalid TYIn in GetMapper: " + typeof(TYIn);
+            _host.Check(typeof(TYIn) == typeof(uint), msg);
+
+            msg = "Invalid TOut in GetMapper: " + typeof(TOut);
+            _host.Check(typeof(TOut) == typeof(float), msg);
+
+            ValueMapper<uint, uint, float> mapper = MapperCore;
+            return mapper as ValueMapper<TXIn, TYIn, TOut>;
+        }
+
+        private void MapperCore(ref uint srcCol, ref uint srcRow, ref float dst)
+        {
+            // REVIEW tfinley: The key-type version a bit more "strict" than the predictor
+            // version, since the predictor version can't know the maximum bound during
+            // training. For higher-than-expected values, the predictor version would return
+            // 0, rather than NaN as we do here. It is in my mind an open question as to what
+            // is actually correct.
+            if (srcRow == 0 || srcRow > _m || srcCol == 0 || srcCol > _n)
+            {
+                dst = float.NaN;
+                return;
+            }
+            dst = Score((int)(srcCol - 1), (int)(srcRow - 1));
+        }
+
+        private float Score(int col, int row)
+        {
+            _host.Assert(0 <= row && row < _m);
+            _host.Assert(0 <= col && col < _n);
+            float score = 0;
+            int poffset = row * _k;
+            int qoffset = col * _k;
+            for (int i = 0; i < _k; i++)
+                score += _p[poffset + i] * _q[qoffset + i];
+            return score;
+        }
+
+        public ISchemaBoundMapper Bind(IHostEnvironment env, RoleMappedSchema schema)
+        {
+            Contracts.AssertValue(env);
+            env.AssertValue(schema);
+            return new RowMapper(this, schema, Schema.Create(new ScoreMapperSchema(OutputType, MetadataUtils.Const.ScoreColumnKind.Regression)));
+        }
+
+        private sealed class RowMapper : ISchemaBoundRowMapper
+        {
+            private readonly MatrixFactorizationPredictor _parent;
+
+            private readonly int _xColIndex;
+            private readonly int _yColIndex;
+            private readonly string _xColName;
+            private readonly string _yColName;
+
+            private IHost Host => _parent._host;
```

Comment:
Feels lazy, why can't you accept env in constructor and create your own host?

Reply:
Fixed. Now we have a private field `_env` assigned by the constructor.